vizro-core/src/vizro/models/_base.py

```diff
@@ -101,8 +108,116 @@ def _is_discriminated_union(field):
         cls.update_forward_refs(**vm.__dict__.copy())
         new_type.update_forward_refs(**vm.__dict__.copy())
 
+    def dict(self, **kwargs):
+        # Overwrite pydantic's own `dict` method to add __vizro_model__ to the dictionary.
+        # To get exclude as an argument is a bit fiddly because this function is called recursively
+        # inside pydantic, which sets exclude=None by default.
+        if kwargs.get("exclude") is None:
+            kwargs["exclude"] = self.__vizro_exclude_fields__()
+        _dict = super().dict(**kwargs)
+        # Inject __vizro_model__ into the dictionary - needed just for to_python.
+        _dict["__vizro_model__"] = self.__class__.__name__
+        return _dict
+
+    # This is like pydantic's own __exclude_fields__ but safer to use (it looks like __exclude_fields__ no longer
+    # exists in pydantic v2).
+    # Root validators with pre=True are always included, even when exclude_default=True, and so this is needed
+    # to potentially exclude fields set this way, like Page.id.
+    def __vizro_exclude_fields__(self) -> Optional[Union[Set[str], Mapping[str, Any]]]:
+        return None
+
+    def _to_python(self, extra_imports: Optional[Set[str]] = None, extra_callable_defs: Set[str] = set()) -> str:
+        """Converts a Vizro model to the Python code that would create it.
+
+        Args:
+            extra_imports: Extra imports to add to the Python code. Provide as a set of complete import strings.
+            extra_callable_defs: Extra callable definitions to add to the Python code. Provide as a set of complete
+                function definitions.
```

Comment:
Should we give examples on how to add `extra_imports` and `extra_callable_defs` here or only in the docs?

Reply:
Currently it is a private function, and it is not even added in the docs. @antonymilne do you think we should:
- make it non-private
- add it to the docs
- announce it in release notes

Any combination of the above?

Reply:
For now I would:
* keep it private
* not should it in docs
* not add to changelog
* incorporate @petar-qb's suggestion of adding example to the docstrings because this would be useful anyway

Once it's used in anger by Vizro-AI and we're definitely happy with it let's then make it public and add to docs and changelog. Just in case e.g. we decide that the current API is actually not convenient for Vizro-AI and we want to change it (shouldn't happen because @maxschulz-COL already thought this through, but still there's no big rush to make it public given that Vizro-AI is the primary consumer to begin with).